### PR TITLE
[exporter/awsemfexporter]Split EMF log with larger than 100 buckets.

### PR DESCRIPTION
**Pull request by zzhlogin**
**Description:** <Describe what has changed.>
In Application Signals, we utilize Base2 Exponential Bucket Histogram to aggregate and send latency data, with a default max number of buckets 160. In EMF exporter, these buckets are mapped to "[Target members](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Embedded_Metric_Format_Specification.html#CloudWatch_Embedded_Metric_Format_Specification_structure_target)" in EMF log entries. 
However, CloudWatch EMF logs impose a limit of 100 target members, beyond which EMF processors will mark the record as `invalid`, resulting in missing metrics and customer-facing errors reported via the EMFValidationErrors metric.

In this PR, we split histograms to two sub EMF logs with the following change:
1. Add an extra attribute `metricIndex` to `groupedMetricMetadata` : Current EMF exporter aggregate incoming metrics into groupedMetrics before converting to log events, where the groupKey is generated based on the groupedMetricMetadata including: metric namespace, timestamp, log group name, etc. After splitting, the two new metrics will share exactly the same key. Adding an extra metric metadata for key generation can prevent the second metric from dropping.
2. If the total buckets exceed 100, the exponential histogram metric are split into into multiple data points as needed,
each containing a maximum of 100 buckets, to comply with CloudWatch EMF log constraints.
For each split data point:
- Min and Max values are recalculated based on the bucket boundary within that specific split.
- Sum is only assigned to the first split to ensure the total sum of the datapoints after aggregation is correct.
- Count is accumulated based on the bucket counts within each split.
 
**Testing:** <Describe what testing was performed and which tests were added.>
The change is tested by generating traffic with more than 100 buckets, and the emf log with larger than 100 values are eliminated after the change:
Screenshot 2024-10-18 at 3 04 45 PM

Compare the added Benchmark test before vs after the code change:
Benchmark test with 100 bucket length:
 ```
goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter
cpu: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
                                                   │ before_100.txt │           after_100.txt            │
                                                   │     sec/op     │   sec/op     vs base               │
GetAndCalculateDeltaDataPointsInclude100Buckets-16      19.68µ ± 5%   20.55µ ± 6%  +4.41% (p=0.015 n=10)

                                                   │ before_100.txt │            after_100.txt            │
                                                   │      B/op      │     B/op      vs base               │
GetAndCalculateDeltaDataPointsInclude100Buckets-16     11.50Ki ± 0%   12.02Ki ± 0%  +4.48% (p=0.000 n=10)

                                                   │ before_100.txt │           after_100.txt           │
                                                   │   allocs/op    │ allocs/op   vs base               │
GetAndCalculateDeltaDataPointsInclude100Buckets-16       126.0 ± 0%   132.0 ± 0%  +4.76% (p=0.000 n=10)

```
Benchmark test with 200 bucket length:
 ```
goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter
cpu: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
                                                   │ before_200.txt │           after_200.txt            │
                                                   │     sec/op     │   sec/op     vs base               │
GetAndCalculateDeltaDataPointsInclude200Buckets-16      26.36µ ± 6%   28.01µ ± 5%  +6.24% (p=0.011 n=10)

                                                   │ before_200.txt │            after_200.txt            │
                                                   │      B/op      │     B/op      vs base               │
GetAndCalculateDeltaDataPointsInclude200Buckets-16     15.50Ki ± 0%   16.59Ki ± 0%  +7.06% (p=0.000 n=10)

                                                   │ before_200.txt │           after_200.txt            │
                                                   │   allocs/op    │ allocs/op   vs base                │
GetAndCalculateDeltaDataPointsInclude200Buckets-16       128.0 ± 0%   152.0 ± 0%  +18.75% (p=0.000 n=10)

```
Benchmark test with 300 bucket length:
 ```
goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter
cpu: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
                                                   │ before_300.txt │           after_300.txt            │
                                                   │     sec/op     │   sec/op     vs base               │
GetAndCalculateDeltaDataPointsInclude300Buckets-16      37.04µ ± 6%   39.17µ ± 5%  +5.73% (p=0.029 n=10)

                                                   │ before_300.txt │            after_300.txt             │
                                                   │      B/op      │     B/op      vs base                │
GetAndCalculateDeltaDataPointsInclude300Buckets-16     23.50Ki ± 0%   20.98Ki ± 0%  -10.70% (p=0.000 n=10)

                                                   │ before_300.txt │           after_300.txt            │
                                                   │   allocs/op    │ allocs/op   vs base                │
GetAndCalculateDeltaDataPointsInclude300Buckets-16       130.0 ± 0%   171.0 ± 0%  +31.54% (p=0.000 n=10)

```
Benchmark test with 500 bucket length:
 ```
goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter
cpu: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
                                                   │ before_500.txt │            after_500.txt            │
                                                   │     sec/op     │   sec/op     vs base                │
GetAndCalculateDeltaDataPointsInclude500Buckets-16      52.51µ ± 3%   58.15µ ± 5%  +10.74% (p=0.000 n=10)

                                                   │ before_500.txt │            after_500.txt             │
                                                   │      B/op      │     B/op      vs base                │
GetAndCalculateDeltaDataPointsInclude500Buckets-16     23.50Ki ± 0%   30.14Ki ± 0%  +28.26% (p=0.000 n=10)

                                                   │ before_500.txt │           after_500.txt            │
                                                   │   allocs/op    │ allocs/op   vs base                │
GetAndCalculateDeltaDataPointsInclude500Buckets-16       130.0 ± 0%   210.0 ± 0%  +61.54% (p=0.000 n=10)

```

### CODE REVIEW

**Review by mxiamxia — APPROVED** [review]

**Review by mxiamxia — COMMENTED** [review]
Please fix the PR checks

**Comment by zzhlogin** [comment]
> This PR was marked stale due to lack of activity. It will be closed in 14 days.

Adding @Aneurysm9 for help on review.

File: exporter/awsemfexporter/datapoint.go [hunk 1]
```diff
@@ -109,6 +109,31 @@ type summaryMetricEntry struct {
 	count uint64
 }
 
+type dataPointSplit struct {
```
[the remark this answers is not in the document]
Comment (zzhlogin) [reply exporter/awsemfexporter/datapoint.go:112]: Each new datapointSplit is created and accessed sequentially in single thread, thus the current design is thread safe. Added comment to note for potential future usage.

File: exporter/awsemfexporter/datapoint.go [hunk 2]
```diff
@@ -193,85 +218,199 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
...
+	if metric.ZeroCount() > 0 {
+		totalBucketLen++
+	}
+
+	for currentBucketIndex < totalBucketLen {
+		// Create a new dataPointSplit with a capacity of up to splitThreshold buckets
+		capacity := splitThreshold
+		if totalBucketLen-currentBucketIndex < splitThreshold {
+			capacity = totalBucketLen - currentBucketIndex
+		}
```
[the remark this answers is not in the document]
Comment (zzhlogin) [reply exporter/awsemfexporter/datapoint.go:248]: Golang does not have a built-in min function. For the suggested change above, it requires creating a separate min function, which introduces unnecessary code complexity for a simple operation. Alternatively, using an inline conditional expression achieves the same result and avoids adding an extra function.

File: exporter/awsemfexporter/datapoint.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Aneurysm9) [reply exporter/awsemfexporter/datapoint.go:113]: Threads aren't really a thing in Go, instead we speak of data as being safe for concurrent use or not.  The comment here should describe what the structure is used for and note that it is not safe for concurrent use.  The user can figure out what type of synchronization is appropriate if they find a need for concurrent use.

**Review by Aneurysm9 — COMMENTED** [review]
Where are the benchmarks?  I don't see them in the PR.  The benchstat output is also missing the after and comparison data.

**Comment by zzhlogin** [comment]
> Where are the benchmarks? I don't see them in the PR. The benchstat output is also missing the after and comparison data.

Sorry, failed to execute the comparison command  before, added the after.txt output in the description now. The benchmark test is located in file "exporter/awsemfexporter/datapoint_test.go" at line 2075, 2076.

File: exporter/awsemfexporter/datapoint.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zzhlogin) [reply exporter/awsemfexporter/datapoint.go:114]: Thanks for the explanation, updated the comment accordingly.

**Review by Aneurysm9 — APPROVED** [review]

File: exporter/awsemfexporter/datapoint.go [hunk 3]
```diff
@@ -193,85 +220,196 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
...
+	scale := metric.Scale()
+	base := math.Pow(2, math.Pow(2, float64(-scale)))
 	negativeBuckets := metric.Negative()
 	negativeOffset := negativeBuckets.Offset()
 	negativeBucketCounts := negativeBuckets.BucketCounts()
-	bucketBegin = 0
-	bucketEnd = 0
-	for i := 0; i < negativeBucketCounts.Len(); i++ {
-		index := i + int(negativeOffset)
+	bucketBegin := 0.0
```
Comment (atoulme) [inline exporter/awsemfexporter/datapoint.go:384]: why not int?

Comment (zzhlogin) [reply exporter/awsemfexporter/datapoint.go:384]: The `bucketBegin` is a float number as it is calculated using: `-math.Pow(base, float64(index+1)),` in previous code, it is also initialized with `var bucketBegin float64`

Comment (atoulme) [reply exporter/awsemfexporter/datapoint.go:383]: I missed that. OK.

**Review by atoulme — APPROVED** [review]

**Review by songy23 — COMMENTED** [review]
This leads to test failures, see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/36727

**Pull request merged**